### ISSUE TITLE
DNAT to LXC seems to not work

### ISSUE THREAD

**Issue by Eloque** [firehol/firehol#144]
I am trying to do something that I would expect to be quite easy, but it seems to be quite hard;
I have a LXC running on guest ip 10.0.3.21 and want to forward SSH traffic on port 8000 to that instance. So I would think

`dnat4 to 10.0.3.21:22 proto tcp dst 83.84.xxx.xxx dport 8000`

And then allowing port 8000 on the internet facing interface would do the trick. But it doesn't. SSH to 10.0.3.21 does work, but to port 8000 on 83.84.xxx.xxx is faced with a connection refused.

I am using the guide at https://firehol.org/guides/firehol-welcome/ the part about DNAT.

**Comment by Eloque** [comment]
I've tried both, to figure out if it was me. But no, that didn't work.

**Comment by philwhineray** [comment]
Tried both what, sorry?

Have you connected in from an external machine? That's the only thing
expected to work.

Is the result of that a connection refused or a timeout? Can you include
your full config and the log output where the source IP is for the server
you are connecting from?

**Comment by Eloque** [comment]
Tried both internal and external. External as only to expected to work. Internal to check if I made a mistake. The result was a connection refused when I tried to SSH from an external IP.

Firehol config is attached. I don't have any logs available, I'll need to look up how to turn those back on.

[firehol.txt](https://github.com/firehol/firehol/files/324199/firehol.txt)

**Comment by philwhineray** [comment]
So the incoming connection is on eth4 and then DNATted to go to a virtual
device correct? What device has the 10.0.3.21 address - I assume it is one
of those matched in the firewall config as veth+?

If so, the DNAT means the packets wind up forwarded from eth4 to that veth#
so you need a router which pairs these. Rules to e.g. allow the bridge to
talk to the internet will not work because the bridge is only involved when
veth devices talk to each other.

You need the logs, ideally, to see what is in each of the fields you are
trying to match.

**Comment by Eloque** [comment]
Correct. That would be vethRGOX1F. I need a seperate router for that?

**Comment by philwhineray** [comment]
I believe so.

Alternately using veth+ and eth4 as inface/outface in a router should work
too.

Try it to confirm, if not we need to get the logs as I don't have any
better ideas right now.

**Comment by Eloque** [comment]
```
router gitlab inface eth4 outface vethRGOX1F
        client all      accept
        server all      accept

```
If I understand correctly? I tried that, but it didn't work. I am enabling logging now, to better analyse.

**Comment by Eloque** [comment]
Re-enabled logging, but nothing shows in the log for my SSH attempts, only other things. Is there a way to force logging of just a specific port or IP?

**Comment by philwhineray** [comment]
You can add: log "my text" to the end of a client or server. You can add a
specific entry in front of your existing ones.

However, firehol logs everything that doesn't match a rule, so chances are
you have a different problem. Likewise the fact it is a connection refused
not a timeout (unless you have specified reject as a default anywhere)
suggests it is not a firehol rule or lack thereof that is blocking you.

Also, I may have been misleading you earlier - does the veth bridge have
its own 10 IP and the vethXXX devices not have IPs in the parent
namespaces? Your existing rules would look OK in that case.

You may also get more information running tcpdump on eth4 and then the veth
bridge to verify you can see the incoming and outgoing packets.

**Comment by Eloque** [comment]
I have "server all reject" as the last statement in the eth4 interface.
Yes, the bridge has it's own ip, the individual devices do not in that way.

As an added complexity, I don't know if that has any consequences, the eth0 device and the tap0 device are bridged in br0 as well, at ip 192.168.40.12, being the LAN. This to allow clients that VPN into the server to behave like full LAN clients.

```
lxcbr0    Link encap:Ethernet  HWaddr fe:24:45:bf:2c:15
          inet addr:10.0.3.1  Bcast:0.0.0.0  Mask:255.255.255.0
          inet6 addr: fe80::fce4:76ff:fef2:20eb/64 Scope:Link
          UP BROADCAST RUNNING MULTICAST  MTU:1500  Metric:1
          RX packets:993 errors:0 dropped:0 overruns:0 frame:0
          TX packets:2080 errors:0 dropped:0 overruns:0 carrier:0
          collisions:0 txqueuelen:0
          RX bytes:144641 (144.6 KB)  TX bytes:429246 (429.2 KB)

vethRGOX1F Link encap:Ethernet  HWaddr fe:24:45:bf:2c:15
          inet6 addr: fe80::fc24:45ff:febf:2c15/64 Scope:Link
          UP BROADCAST RUNNING MULTICAST  MTU:1500  Metric:1
          RX packets:993 errors:0 dropped:0 overruns:0 frame:0
          TX packets:2074 errors:0 dropped:0 overruns:0 carrier:0
          collisions:0 txqueuelen:1000
          RX bytes:158543 (158.5 KB)  TX bytes:427576 (427.5 KB)
```

**Comment by philwhineray** [comment]
Ah, I missed the "server all reject".

If this line is being hit (you can add a "log whatever" to it to see the
packets explicitly), that would explain the connection refused.

I notice now that the dnat is after the eth4 interface - it should probably
be before in general. Otherwise it will behave as if it is "in" the
interface and I'm not sure how those semantics work in relation to the NAT
tables.

**Comment by Eloque** [comment]
So, before the eth4 and I can use the log bit. Did try that now. No success, no log. However, if I try the SSH command to a port I know that isn't available, I do get a log, as it is being rejected. 

So, if I create the DNAT rule, the packet isn't dropped or rejected. But if I do, something else is refusing I believe. I put logging on the server accept line, and this is what I get;

```
Jun 21 20:29:38 ruby kernel: [848443.704876] Gitlab_accept:IN=eth4 OUT= MAC=bc:5f:f4:1c:90:b4:00:01:5c:6c:7c:46:08:00 SRC=198.46.xxx.xxx DST=83.84.xxx.xxx LEN=60 TOS=0x00 PREC=0x00 TTL=54 ID=46654 DF PROTO=TCP SPT=52355 DPT=802 WINDOW=14600 RES=0x00 SYN URGP=0
Jun 21 20:29:38 ruby kernel: [848443.704997] Gitlab_accept:IN= OUT=eth4 SRC=83.84.xxx.xxx DST=198.46.xxx.xxx LEN=40 TOS=0x00 PREC=0x00 TTL=64 ID=12444 DF PROTO=TCP SPT=802 DPT=52355 WINDOW=0 RES=0x00 ACK RST URGP=0
```
I am thinking that a router is missing? I do have this;
```
router gitlab inface eth4 outface vethRGOX1F
        client all      accept
        server all      accept

```

**Comment by philwhineray** [comment]
Can I ask exactly what firehol version you have?

I have just tried running your config for the first time, to see exactly
what statements are created and it fails, telling me that the dnat command
must come before all interfaces.

Notwithstanding that, if the dnat is working, then the interface definition
should not be being used. I don't think you need the extra router because
the routing is done from eth4 to the bridge device. There should be a route
loaded in the routing table which says anything for the 10 network goes
there.

FireHOL only logs by default if no rule is being matched, so...

If the dnat is not matching, then then server all reject in the interface
is matched and no log is emitted (unless you tell it to). Another way to
test this will be to set the policy to reject and remove the "server all
reject" temporarliy... then you should get the log.

Another possibility is that the dnat is working as expected. If so, there
is no log because a "server all accept" is being hit in the appropriate
router. If this is the case, the container is rejecting the connection for
its own reasons - e.g. it can have its own iptables setup loaded. If that
is the case, I would expect to be able to run tcpdump on the bridge and see
the connection and the response.

You do not have to specify any interfaces in a firehol router, by the way,
any FORWARD traffic tries each router in turn by default, so this is legal,
and essentially the same as you have:

# Put these at the top
masquerade eth4
masquerade lxcbr0

dnat4 to 10.0.3.21:22 proto tcp dst 83.84.1.1 dport 8000

#
## Your interface definitions here
#

router allroutes
        server all      accept
        client all      accept

**Comment by Eloque** [comment]
FireHOL 3.0.2_master is the version I am using. Building a test conf file based on what you just send does work. Adjusting my main configuration after that based on that, does also work.

Am I understanding correctly that I can replace all the routers with that single router statement and it will still work and be safe;

```
router allroutes
        server all      accept
        client all      accept
```

**Comment by philwhineray** [comment]
Well, it's pretty much equivalent to what you had before.

I'm not sure that having a router accepting all traffic in all directions can be considered safe. You should probably replace the rules with more specific ones. The rules you have on the interface protect the host but only the host, which has always been the case.

If you add logging to the client and server lines in allroutes, you can find out what the inface and outface values are if you want to make per-case routers again.
